### PR TITLE
AWS: SSL support for ELB listeners through annotations

### PR DESCRIPTION
**Pull request by therc**
In the API, ports have only either TCP or UDP as their protocols, but ELB distinguishes HTTPS->HTTP[S]? from SSL->(SSL|TCP). 

Per #24978, this is implemented through two separate annotations:

`service.beta.kubernetes.io/aws-load-balancer-ssl-cert=arn:aws:acm:us-east-1:123456789012:certificate/12345678-1234-1234-1234-123456789012`
`service.beta.kubernetes.io/aws-load-balancer-backend-protocol=(https|http|ssl|tcp)`

Mixing plain-text and encrypted listeners will be in a separate PR, implementing #24978's `aws-load-balancer-ssl-ports=LIST`

### CODE REVIEW

**Comment by therc** [comment]
cc @justinsb

**Comment by therc** [comment]
I'm working on tests, but I thought I'd submit the new code before the old one gets looked at. I've implemented the `protocol=certARN` syntax, added checks and errors, moved the listener creation to an earlier stage so that we don't leave security groups behind and added comments/pointers.

**Comment by therc** [comment]
I added tests. I gave up on trying FakeELB-based tests: I simply added tests for the new getListener() function (another good reason to factor it out of EnsureLoadBalancer).

**Comment by justinsb** [comment]
Nice.  A few thoughts:

1. Is HTTP vs TCP orthogonal here? e.g. should it maybe be a different annotation.  There is also another WIP PR (#24569) to add PROXY protocol, so I'd like to figure out what our final set of annotations should be (once we have SSL, PROXY, HTTP etc).
1. Do we want to support http & https?  Even if the http listener is only going to redirect to https, I think we still need it.  But I'm also not sure exactly how we solve this cleanly within the framework we have.

Random thoughts (no action required):

1. I'm wondering if we should use the arn, or use tags to create something easier to manage.  But if you're requiring a full arn e.g. `arn:aws:iam::123456789012:server-certificate/company/servercerts/ProdServerCert` then that works fine because we can always look at the `arn:` prefix.
1. A long-term goal of k8s (as I understand it) is to allow multiple untrusted tenants to share a k8s cluster.    So being able to use any cert you know the name of is not ideal.  But I imagine we would solve this with an ACL tag on the cert, or an ACL entry when we implement real ACLs.

**Comment by therc** [comment]
I'm wary of needing multiple annotations (cert, http/tcp) just for the SSL certificate. I do agree we just keep all of the ELB annotations in sync.

As for http, maybe we can have a prefix or suffix indicating the creation of a plain text listener, too. I think our default use case should be HTTPS/SSL only. Redirects are usually a bad idea, see e.g. http://www.vinaysahni.com/best-practices-for-a-pragmatic-restful-api#ssl

I think that in the long run an Ingress with all the necessary checks is the way to go for multiple tenancy.

**Comment by dwelch2344** [comment]
Hey all, just weighing in... 

```@justinb: Do we want to support http & https? Even if the http listener is only going to redirect to https, I think we still need it. But I'm also not sure exactly how we solve this cleanly within the framework we have.```
This is a definite must IMO. Doesn't necessarily need to be there first cut, but I would say that a lot of people (including myself) would expect this ability and get burned without it.

Using tags or other annotations are totally fine IMO, even if it's a little messy in that you need to wire a few of them together to make it work.

**Comment by therc** [comment]
I split the single annotation into `aws-load-balancer-ssl-cert` / `aws-load-balancer-backend-protocol=(https|http|ssl|tcp)` and updated the PR description. Mixed listeners will be implemented separately.

**Comment by therc** [comment]
@justinsb I think this is ready for review. I'll implement #24978's `aws-load-balancer-ssl-ports=LIST` as soon as this one lands.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -69,6 +69,26 @@ const TagNameSubnetPublicELB = "kubernetes.io/role/elb"
...
+// Service annotation requesting a secure listener. Value is a valid certificate ARN.
+// For more, see http://docs.aws.amazon.com/ElasticLoadBalancing/latest/DeveloperGuide/elb-listener-config.html
+// CertARN is an IAM or CM certificate ARN, e.g. arn:aws:acm:us-east-1:123456789012:certificate/12345678-1234-1234-1234-123456789012
+const ServiceAnnotationLoadBalancerCertificate = "service.beta.kubernetes.io/aws-load-balancer-ssl-cert"
+
+// Service annotation specifying the protocol spoken by the backend (pod) behind a secure listener.
+// Only inspected when `aws-load-balancer-ssl-cert` is used.
+// If `http` (default) or `https`, an HTTPS listener that terminates the connection and parses headers is created.
+// If set to `ssl` or `tcp`, a "raw" SSL listener is used.
+
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Nit: remove blank line here

File: pkg/cloudprovider/providers/aws/aws.go [hunk 2]
```diff
@@ -2099,6 +2119,35 @@ func isSubnetPublic(rt []*ec2.RouteTable, subnetID string) (bool, error) {
 	return false, nil
 }
 
+func getListener(port api.ServicePort, annotations map[string]string) (*elb.Listener, error) {
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Nit: maybe call this buildListener?  That's more what it is doing, right?  (A comment might be helpful too!)

File: pkg/cloudprovider/providers/aws/aws.go [hunk 3]
```diff
@@ -2099,6 +2119,35 @@ func isSubnetPublic(rt []*ec2.RouteTable, subnetID string) (bool, error) {
...
+	instanceProtocol := protocol
+
+	listener := &elb.Listener{}
+	listener.InstancePort = &instancePort
+	listener.LoadBalancerPort = &loadBalancerPort
+	certID := annotations[ServiceAnnotationLoadBalancerCertificate]
+	if certID != "" {
+		instanceProtocol = annotations[ServiceAnnotationLoadBalancerBEProtocol]
+		if instanceProtocol == "" {
+			protocol = "https"
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Hmm... wondering if we should default to HTTP or TCP.  If we don't have ServiceAnnotationLoadBalancerCertificate, the default is TCP.  I think on that grounds it would be surprising if you set ServiceAnnotationLoadBalancerCertificate and then the default switched to HTTP.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 4]
```diff
@@ -2099,6 +2119,35 @@ func isSubnetPublic(rt []*ec2.RouteTable, subnetID string) (bool, error) {
...
+			instanceProtocol = "http"
+		} else {
+			protocol = backendProtocolMapping[instanceProtocol]
+			if protocol == "" {
+				return nil, fmt.Errorf("Invalid backend protocol %s in %s", instanceProtocol, certID)
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Maybe reference the name of the annotation, to give users more of a hint here?

File: pkg/cloudprovider/providers/aws/aws.go (hunk 2 above)
Comment (therc) [inline pkg/cloudprovider/providers/aws/aws.go]: Done both

File: pkg/cloudprovider/providers/aws/aws.go (hunk 1 above)
Comment (therc) [inline pkg/cloudprovider/providers/aws/aws.go]: Done

**Comment by justinsb** [comment]
A few nits which you can fix or not as you see fit.

The default value for protocol though I think is more important... I think it would be surprising to set the ssl-cert annotation and have the protocol switch to HTTP.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 3 above)
Comment (therc) [inline pkg/cloudprovider/providers/aws/aws.go]: With HTTP, users get the X-Forwarded-For header. But I guess it would be surprising. Changed and added a test (I had all cases except the default one...).

File: pkg/cloudprovider/providers/aws/aws.go (hunk 4 above)
Comment (therc) [inline pkg/cloudprovider/providers/aws/aws.go]: Done

**Comment by therc** [comment]
I fixed all nits and changed the default.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 5]
```diff
@@ -2099,6 +2118,37 @@ func isSubnetPublic(rt []*ec2.RouteTable, subnetID string) (bool, error) {
...
+	instanceProtocol := protocol
+
+	listener := &elb.Listener{}
+	listener.InstancePort = &instancePort
+	listener.LoadBalancerPort = &loadBalancerPort
+	certID := annotations[ServiceAnnotationLoadBalancerCertificate]
+	if certID != "" {
+		instanceProtocol = annotations[ServiceAnnotationLoadBalancerBEProtocol]
+		if instanceProtocol == "" {
+			protocol = "ssl"
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Wondering if we should warn/error in the case when the user specifies an invalid annotation (i.e. not found in map, vs not found).  But that can be a separate PR - I want to get this one merged!

Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Edit: never mind... I need more coffee

**Comment by justinsb** [comment]
LGTM :-)

**Comment by dkozma** [comment]
@therc This looks great - thanks for your work on this. When testing this out, I tried to set up the following:

ELB HTTPS/:443 -> Service HTTP listener port
ELB HTTP/:80 -> Service HTTP listener port

by setting:

```yml
metadata:
    annotations:
        service.beta.kubernetes.io/aws-load-balancer-ssl-cert: arn:aws:acm:us-east-1:xxxxx:certificate/xxxxx
        service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
spec:
  type: LoadBalancer
  ports:
    - port: 80
      targetPort: 8000
      protocol: TCP
      name: http
    - port: 443
      targetPort: 8000
      protocol: TCP
      name: https
```

From this setup, I got:

ELB HTTPS/:443 -> HTTP
ELB HTTP**S**/:80 -> HTTP

Is this expected behavior? If not, is there a way this can be set up to allow HTTP connections over 80 and not HTTPS?

**Comment by therc** [comment]
@dkozma that's expected, but #26268 will take care of your case.

**Comment by dkozma** [comment]
@therc Perfect, thanks!

**Comment by therc** [comment]
@dkozma would you be able to test #26976?

**Comment by dkozma** [comment]
@therc I can spin up a cluster tomorrow and let you know the results.

**Comment by dkozma** [comment]
@therc Sorry for the delay on this. Just built your branch and spun up a new cluster and everything works great - I have ELB listeners for 80 and 443 working correctly.. Thank you for your efforts in adding this feature.

**Comment by razic** [comment]
@therc 
@dkozma 

I have a suggestion to this feature. Currently the annotation is named `aws-load-balancer-backend-protocol`. However, look at the image below:

screen shot 2016-08-08 at 5 12 48 pm

AWS uses the terminology `instance` and not `backend`. To avoid confusion, and to ensure consistency, I suggest that we change

```
aws-load-balancer-backend-protocol
```

to 

```
aws-load-balancer-instance-protocol
```

Please let me know your thoughts. Thanks!

**Comment by quinton-hoole** [comment]
cc @kubernetes/sig-federation-misc
